### PR TITLE
Implement PIN-based two-factor authentication

### PR DESCRIPTION
**Pull request by Chaosvex**
_Update: TOTP ported._
_Note: This PR is currently for reviewing/feedback rather than merging._

This is an implementation of the two-factor PIN system that's available in the 1.xx clients. It should also be available in TBC, WotLK and possibly beyond.

I've ported it directly from my own original implementation (https://github.com/EmberEmu/Ember/blob/spark-new/src/login/PINAuthenticator.cpp)/

So you can see how it behaves without running it yourself:
https://www.youtube.com/watch?v=_9IS44BGgJU      - Fixed PIN
https://www.youtube.com/watch?v=VXAOxSBGFXg   - TOTP

I'll squash the commits before the final PR. :)

Thanks!

### CODE REVIEW

File: sql/base/realmd.sql [hunk 1]
```diff
@@ -1,8 +1,8 @@
--- MySQL dump 10.13
+-- MySQL dump 10.13  Distrib 5.6.17, for Win64 (x86_64)
```
Comment (Phatcat) [inline sql/base/realmd.sql]: This

File: sql/base/realmd.sql [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 --
--- Host: localhost    Database: realmd
+-- Host: 127.0.0.1    Database: realmd3
```
Comment (Phatcat) [inline sql/base/realmd.sql]: this

File: sql/base/realmd.sql [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
 -- ------------------------------------------------------
--- Server version	5.5.32
+-- Server version 5.6.21
```
Comment (Phatcat) [inline sql/base/realmd.sql]: and this, probably has to go.

File: sql/base/realmd.sql [hunk 4]
```diff
@@ -16,30 +16,12 @@
 /*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;
 
 --
--- Table structure for table `realmd_db_version`
```
Comment (Phatcat) [inline sql/base/realmd.sql]: Why did you remove all of this?

File: sql/base/realmd.sql [hunk 5]
```diff
@@ -16,30 +16,12 @@
...
-(NULL);
-/*!40000 ALTER TABLE `realmd_db_version` ENABLE KEYS */;
-UNLOCK TABLES;
-
---
 -- Table structure for table `account`
 --
 
 DROP TABLE IF EXISTS `account`;
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
```
Comment (Phatcat) [inline sql/base/realmd.sql]: you've added a lot of these kind of lines which are prety much all commented out ?

File: src/realmd/AuthSocket.cpp [hunk 6]
```diff
@@ -511,6 +513,13 @@ bool AuthSocket::_HandleLogonProof()
     if (!Read((char*)&lp, sizeof(sAuthLogonProof_C)))
         return false;
 
+    PINData pinData;
+    if(lp.securityFlags)
+    {
+        if(!Read((char*)&pinData, sizeof(pinData)))
+            return false;
+    }
+
```
Comment (Phatcat) [inline src/realmd/AuthSocket.cpp]: could maybe simplify into;

`if (lp.securityFlags && !Read((char*)&pinData, sizeof(pinData)))`
`    return false;`

?

File: src/realmd/AuthSocket.cpp [hunk 7]
```diff
@@ -511,6 +513,13 @@ bool AuthSocket::_HandleLogonProof()
     if (!Read((char*)&lp, sizeof(sAuthLogonProof_C)))
         return false;
 
+    PINData pinData;
```
Comment (Phatcat) [inline src/realmd/AuthSocket.cpp]: maybe add an empty newline between this and the line below?

Comment (Chaosvex) [inline src/realmd/AuthSocket.cpp]: That'd be my preference but I just followed the example set above by "sAuthLogonProof_C lp;".

File: sql/base/realmd.sql (hunk 5 above)
Comment (Chaosvex) [inline sql/base/realmd.sql]: It's an autogenerated file.

File: sql/base/realmd.sql (hunk 4 above)
Comment (Chaosvex) [inline sql/base/realmd.sql]: It's an autogenerated file. I guess the exporter changed slightly since the table was last updated in 2013. ;)

File: sql/base/realmd.sql (hunk 3 above)
Comment (Chaosvex) [inline sql/base/realmd.sql]: These details don't really matter. As said elsewhere, autogenerated file. I'll change the database name at least before it's final. ;)

File: src/realmd/AuthSocket.cpp (hunk 7 above)
Comment (Phatcat) [inline src/realmd/AuthSocket.cpp]: I'd say add an empty line to the example above as well, this is not one of those files being several thousand lines long so we don't have to feel bad using a bit of space to keeping it better visually organized in my opinion.

File: src/realmd/AuthSocket.cpp (hunk 6 above)
Comment (Phatcat) [inline src/realmd/AuthSocket.cpp]: no a fan of the idea of turning the two if statements into a single one? or not possible within this scope?

Comment (Chaosvex) [inline src/realmd/AuthSocket.cpp]: It's possible although I think it decreases readability _slightly_.

Comment (Phatcat) [inline src/realmd/AuthSocket.cpp]: Not if you slap a comment on there as well ;)
